### PR TITLE
Narrow down Datatypes

### PR DESCRIPTION
**Pull request by cocker-cc**
- use Datatypes from Stdlib
- use Variant[Integer, Pattern[/^[0-9]+$/]] where Strings represent Integers

### CODE REVIEW

File: manifests/balancer.pp
Comment (ekohl) [inline manifests/balancer.pp:46]: I'd say we shouldn't introduce alignment. With data types it only makes things very messy. It creates a lot of merge conflicts, even for trivial PRs when a change is made.

File: manifests/balancermember.pp
Comment (ekohl) [inline manifests/balancermember.pp:43]: Just for context, https://httpd.apache.org/docs/2.4/mod/mod_proxy.html#balancermember is the documentation. That just mentions url, but I think HTTP/HTTPS is sufficient in schemes.

File: manifests/balancermember.pp
Comment (ekohl) [inline manifests/balancermember.pp:44]: Given the other part is a key value, I wonder why options is an array and not a hash but that's probably outside of the scope. The alternative is to make it `Array[Pattern[/=/]]` just to narrow it down to something valid.

File: manifests/fastcgi/server.pp
Comment (ekohl) [inline manifests/fastcgi/server.pp:39]: This isn't really an absolute path, it's the path in the URL. That's probably closed to an absolute unix path. It shouldn't accept a Windows path though.

File: manifests/mod.pp
Comment (ekohl) [inline manifests/mod.pp:36]: This file only contains whitespace changes? It would be helpful if that would be split off in its own commit.

File: manifests/mod/cluster.pp
Comment (ekohl) [inline manifests/mod/cluster.pp:57]: Given the next version is a major version bump, perhaps we should just go for a breaking change and only allow Stdlib::Port?

File: manifests/mod/authn_dbd.pp
Comment (ekohl) [inline manifests/mod/authn_dbd.pp:37]: I'd consider a breaking change and make it `Integer[0]` (assuming 0 really is the lower bound).

**Review by ekohl — COMMENTED** [review]
I am very happy with stricter data types, but this is quite hard to review due to the various things going on. I'd prefer starting with the simple things like whitespace/coding style changes. Those can be merged pretty easily.

File: manifests/balancer.pp
Comment (cocker-cc) [reply manifests/balancer.pp:46]: I don't care, which Style.  But the Style should be consistent throughout the Module.  
So, what's your Suggestion?

File: manifests/balancermember.pp
Comment (cocker-cc) [reply manifests/balancermember.pp:43]: Just as an additional Note:  `stdlib/types/httpurl.pp`:  
```
type Stdlib::HTTPUrl = Pattern[/(?i:\Ahttps?:\/\/.*\z)/]
```
So, HTTPS is included there.

File: manifests/fastcgi/server.pp
Comment (cocker-cc) [reply manifests/fastcgi/server.pp:39]: I changed to `Stdlib::Unixpath` at several Code-Locations. (next Push)

File: manifests/mod.pp
Comment (cocker-cc) [reply manifests/mod.pp:36]: reverted (next Push)

File: manifests/mod/cluster.pp
Comment (cocker-cc) [reply manifests/mod/cluster.pp:57]: I strongly vote for this Option.  Give me a „go“, and I will narrow down to `Stdlib::Port`.

File: manifests/mod/authn_dbd.pp
Comment (cocker-cc) [reply manifests/mod/authn_dbd.pp:37]: Neither do I know the correct Integer-Ranges.  Possibly there is a `-1` allowed somewhere?  I don't know.  
But I am with you, removing the `Pattern`.  Give me a „go“.

File: manifests/balancer.pp
Comment (ekohl) [reply manifests/balancer.pp:46]: I agree it should be consistent and only later noticed that other files do appear to use alignment. I'd start with a PR that only does this: stylistic changes.

@david22swan thoughts on which style to use? Perhaps that's also something all puppetlabs modules should adopt. Maybe even adopt in https://github.com/puppetlabs/language-style-guide as a recommended way for all module authors.

File: manifests/balancermember.pp
Comment (ekohl) [reply manifests/balancermember.pp:43]: Yes, I was mostly thinking out loud. Should have been explicit in that.

File: manifests/balancer.pp
Comment (david22swan) [reply manifests/balancer.pp:46]: From personal experience I would say I tend towards aligning variables, especially when there are a larger amount of variables, it was something that I was taught early on in Puppet and have always found it easier to read.

Comment (david22swan) [reply manifests/balancer.pp:46]: As to being part of the style guide, it actually is:
https://github.com/puppetlabs/language-style-guide/blob/main/style_guide.md#parameter-indentation-and-alignment
With the instruction being that the equals signs should be aligned.

Comment (cocker-cc) [reply manifests/balancer.pp:46]: I throw into the Pit: `pdk validate`,  which calls `puppet-lint`, which implements the Style-Guide.

Comment (ekohl) [reply manifests/balancer.pp:46]: It's actually something I disagree with. Once I get back from leave I'll see about raising this in the chats we've been having about updating the style guide.

**Comment by david22swan** [comment]
@cocker-cc Look's like a lot of good changes so far.
Looking forward to getting this merged.

**Comment by cocker-cc** [comment]
> @cocker-cc Look's like a lot of good changes so far. Looking forward to getting this merged.

@david22swan: just wait for the Changes out of the Discussions above.  Okay?

**Comment by david22swan** [comment]
> > @cocker-cc Look's like a lot of good changes so far. Looking forward to getting this merged.
> 
> @david22swan: just wait for the Changes out of the Discussions above. Okay?

Yeh, that's what I meant sorry. Should have been clearer.

**Comment by david22swan** [comment]
@cocker-cc Hey, just checking if there are any updates on this coming soon?

File: manifests/mod/authn_dbd.pp
Comment (ekohl) [reply manifests/mod/authn_dbd.pp:37]: My point was that if it's just numbers then reduce it to just making it `Integer`, possibly with a lower bound. We're doing a breaking change anyway so it's ok to be backwards incompatible.

File: manifests/mod/info.pp
Comment (ekohl) [inline manifests/mod/info.pp:23]: This should probably be a `Stdlib::Unixpath`.

**Review by ekohl — CHANGES_REQUESTED** [review]
Please rebase this now that the related PRs have been merged.

Also, as I suggested inline, feel free to narrow things down if they make more sense. It's going to be a major version bump, let's make use of it.

File: manifests/init.pp
Comment (cocker-cc) [reply manifests/init.pp]: I created a Type: `Apache::ServerTokens`

File: manifests/init.pp
Comment (cocker-cc) [reply manifests/init.pp]: I added `extended`

File: manifests/vhost.pp
Comment (cocker-cc) [reply manifests/vhost.pp]: changed

File: manifests/mod/negotiation.pp
Comment (ekohl) [inline manifests/mod/negotiation.pp:20]: Linting fails on this which breaks CI.

File: manifests/init.pp
Comment (ekohl) [inline manifests/init.pp:484]: You could reuse a type
```suggestion
  Variant[Boolean, Stdlib::Ensure::Service] $service_ensure               = 'running',
```

I'd also consider dropping the `Boolean`.

**Review by ekohl — COMMENTED** [review]
Once CI passes I'd be happy to merge this and iterate on it further if we find breakages.

File: manifests/mod/negotiation.pp
Comment (cocker-cc) [reply manifests/mod/negotiation.pp]: I restored the initial Version from `main`.

File: manifests/init.pp
Comment (cocker-cc) [reply manifests/init.pp]: 👍

**Review by ekohl — DISMISSED** [review]
I'm going to merge this. We'll probably see some fallout from this given its size, but it's moving things forward a lot.

**Comment by ekohl** [comment]
Oh, I missed that there is now a merge conflict in `spec/acceptance/mod_php_spec.rb`. Once you fix that, I'll merge this soon to avoid further conflicts.

**Comment by cocker-cc** [comment]
> Oh, I missed that there is now a merge conflict in `spec/acceptance/mod_php_spec.rb`. 

done

File: spec/classes/mod/ldap_spec.rb
Comment (ekohl) [inline spec/classes/mod/ldap_spec.rb:31]: Rubocop is unhappy about this and wants this
```suggestion
          ldap_shared_cache_size: 500_000,
```

Comment (cocker-cc) [reply spec/classes/mod/ldap_spec.rb]: done

**Review by ekohl — APPROVED** [review]
Let's see if CI is green.

**Comment by cocker-cc** [comment]
> Error while evaluating a Resource Statement

I updated some Puppet-Datatypes.

**Comment by cocker-cc** [comment]
Again I reverted some Datatypes back to `String`.

**Comment by cocker-cc** [comment]
```
2022-07-27T14:20:28.5731650Z        expected that the catalogue would contain File[30-rspec.example.com.conf symlink] with path set to "/etc/apache2/sites-enabled/30-rspec.example.com.conf" but it is set to "/etc/httpd/sites-enabled/30-rspec.example.com.conf"
```

Have no Idea, why these appear.  In the [Specfile](https://github.com/cocker-cc/puppetlabs-apache/blob/Narrow_down_Datatypes/spec/defines/vhost_spec.rb#L12) it's set to `apache2`, and in the [PP-File](https://github.com/cocker-cc/puppetlabs-apache/blob/Narrow_down_Datatypes/manifests/params.pp#L332) it's set to `apache2`.

**Review by ekohl — COMMENTED** [review]
Those also happen in main. I'll take a quick look.

**Comment by ekohl** [comment]
I believe https://github.com/puppetlabs/puppetlabs-apache/pull/2266 should fix it.

**Comment by ekohl** [comment]
Looks like the tests will be green (Puppet 7 already passed). The only thing appears to be the CLA.

**Review by david22swan — APPROVED** [review]
It's been a long road but LGTM
Thank you @cocker-cc for putting in the work and @ekohl for all the reviews.

Won't merge immediately as with the size of this I'd like some others on the team to give it a look over first, but all being good should be merged in by Monday evening at the latest, with it and all the other changes released by the next Monday.

File: spec/classes/mod/ldap_spec.rb
Comment (pmcmaw) [inline spec/classes/mod/ldap_spec.rb:31]: Is the underscore a typo?

Comment (ekohl) [reply spec/classes/mod/ldap_spec.rb:31]: No, in Ruby this is a convention to make numbers easier to read. Because it's quite hard to see if it's 5 or 6 zeros, this makes it visually easier to see it's 500k, not 5m. Rubocop even enforces it.

**Comment by pmcmaw** [comment]
I have just posted one of the tiniest questions, other than that it LGTM.

File: spec/classes/mod/ldap_spec.rb
Comment (pmcmaw) [reply spec/classes/mod/ldap_spec.rb:31]: Amazing, never knew this 👍 will definitely come in handy at some point :-)

**Review by pmcmaw — APPROVED** [review]

**Comment by david22swan** [comment]
@cocker-cc Could you sign the CLA, once done I can just hit merge

**Comment by cocker-cc** [comment]
> @cocker-cc Could you sign the CLA? Once that is done I can hit merge

I did, but it does not show up.  

Screenshot from 2022-07-29 12-12-43

File: spec/classes/mod/ldap_spec.rb
Comment (cocker-cc) [reply spec/classes/mod/ldap_spec.rb:31]: @pmcmaw: Sharp Eyes 👀  
I do not like it, but Rubocop with its Config-File coming from the PDK-Templates sais so.

**Comment by david22swan** [comment]
Yeh sorry bout that, it can happen sometimes.
Screenshot's good enough for me, will go ahead and merge.
Thank's again to the both of you for putting in all this work :)

**Comment by pmcmaw** [comment]
@cocker-cc Amazing PR, thank you for your time and effort! 🥳

**Pull request merged**
